File: src/i1tbfssymmetric/Java/com/example/tbfssymmetric/SymmetricTreeChecker.java
package com.example.tbfssymmetric;

public class SymmetricTreeChecker {

    public static boolean isSymmetric(TreeNode<Integer> root) {
        if (root == null)
            return true;

        CustomQueue<TreeNode<Integer>> queue = new CustomQueue<>();
        queue.enqueue(root.left);
        queue.enqueue(root.right);

        while (queue.size() > 0) {
            TreeNode<Integer> left = queue.dequeue();
            TreeNode<Integer> right = queue.dequeue();

            if (left == null && right == null)
                continue;
            if (left == null || right == null)
                return false;
            if (!left.data.equals(right.data))
                return false;

            queue.enqueue(left.left);
            queue.enqueue(right.right);
            queue.enqueue(left.right);
            queue.enqueue(right.left);
        }

        return true;
    }
}

File: src/i1tbfssymmetric/CS/TBFSSymmetric.Algo/SymmetricTreeChecker.cs
namespace TBFSSymmetric.Algo;

public static class SymmetricTreeChecker
{
    public static bool IsSymmetric(TreeNode<int>? root)
    {
        if (root == null)
        {
            return true;
        }

        CustomQueue<TreeNode<int>?>? queue = new();
        queue.Enqueue(root.Left);
        queue.Enqueue(root.Right);

        while (queue.Size > 0)
        {
            TreeNode<int>? left = queue.Dequeue();
            TreeNode<int>? right = queue.Dequeue();

            if (left == null && right == null)
            {
                continue;
            }

            if (left == null || right == null)
            {
                return false;
            }

            if (!left.Data.Equals(right.Data))
            {
                return false;
            }

            queue.Enqueue(left.Left);
            queue.Enqueue(right.Right);
            queue.Enqueue(left.Right);
            queue.Enqueue(right.Left);
        }

        return true;
    }
}
